rfcs/react-components/convergence/global-contex-registration.md

```diff
@@ -0,0 +1,148 @@
+    if (Symbol.keyFor(sym)?.startsWith(SYMBOL_NAMESPACE)) {
+      console.log('deleting', sym);
+      delete globalObject[sym];
+    }
+  });
+}
+
+/**
+ * Wrapper around @see React.createContext that implements context registration
+ * in the globalThis object to avoid duplicate contexts. Contexts are keyed with
+ * a unique sybmol for the package name, version and name of the context.
+ *
+ * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol}
+ *
+ * @param defaultValue - @see React.createContext
+ * @param name - name of the context
+ * @param packageName - name of the npm package where the module is used
+ * @param packageVersion - version of the npm package where the module is used
+ * @returns @see React.createContext
+ */
+export const createContext = <T>(defaultValue: T, name: string, packageName: string, packageVersion: string) => {
+  // Symbol guaranteed to be unique for the entire runtime
+  const sym = Symbol.for(`${SYMBOL_NAMESPACE}${packageName}/${name}/@${major(packageVersion)}`);
+
+  // Objects keyed with symbols are not visible with console.log
+  // Object symbol properties can't be iterated with `for` or `Object.keys`
+  const globalSymbols = Object.getOwnPropertySymbols(globalObject);
+  if (!globalSymbols.includes(sym)) {
+    globalObject[sym] = React.createContext(defaultValue);
+  }
+
+  return globalObject[sym] as React.Context<T>;
+};
+```
+
+The proposed solution involves a wrapper around `React.createContext` which also uses a user defined name, package name
+and package major version as a key to create a symbol on the global object.
+
+When using this solution, the underlying javascript is not actually typed, so any backwards compatible context values
+will work. However, if the context value change is breaking (e.g. removing a property, changing from string to object)
+then the package containing the context should be major bumped.
+
+The solution will result in one single context that is used for each major version.
+
+[Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)
+is used as the key for global contexts, since symbol properties on objects are only visible through the
+`Object.getOwnPropertySymbols` API, this will reduce the possibility that developers will access these global
+contexts to abuse them. `Symbol` is quite an old APi that is supported from Chrome version 40 and Firefox version 36.
+
+Although not in the prototypes, it should be possible to create a global flag that will enable this behaviour in
+Fluent, which means that this feature **should be completely opt-in**, and only intended for partners with complex
+apps and dependency chains.
+
+### Pros and Cons
+
+#### Cons
+
+- Polluting the global namespace
+- Custom magic when creating contexts in Fluent
+- Generally using window/node globals is an antipattern
+
+#### Pros
+
+- One single major version context regardless of duplicate dependencies
```

Comment:
It is perhaps obvious but I would still explicitly mention that with this approach it will be our responsibility to provide backward and forward compatibility of the context values among minor versions.

Reply:
you are right, I will state this and add some examples of traps for forwards/backwards compat

Reply:
Added in [0ddb757](https://github.com/microsoft/fluentui/pull/22300/commits/0ddb7578fcaa55600e600d4c27c142ab50882912)